### ISSUE TITLE
Feature request: add `additionalFormParams` to `authConfigs`

### ISSUE THREAD

**Issue by Kizmar** [swagger-api/swagger-ui#5399]
In [oauth2-authorize.js](https://github.com/swagger-api/swagger-ui/blob/1e184e8e218676278c83e60a45846c199ce3d15e/src/core/oauth2-authorize.js#L69) the `additionalQueryStringParams` property is used from the `authConfigs` object in order to add custom query string parameters to an authentication request.

Could we get a similar `additionalFormParams` property added that could then be passed to the `authActions` via method params or a property? That would allow [actions.js](https://github.com/swagger-api/swagger-ui/blob/937c8f6208f3adf713b10a349a82a1b129bd0ffd/src/core/plugins/auth/actions.js#L124) to add custom form properties as it's building the form data string.

**Reason:**
We're trying to use Swagger-UI (via [Swashbuckle.AspNetCore](https://github.com/domaindrivendev/Swashbuckle.AspNetCore)) to authenticate with Auth0. It has the right mix of params for the `implicit` grant type, but we're finding that `client_credentials` and `password` out-of-the-box body params passed aren't meeting Auth0's requirements. For example: when trying to use `client_credentials`, Auth0 requires `audience` in the body, but swagger-ui only passes `grant_type` and `scope`. Adding `audience` to the query string params doesn't cut it. This is where people are talking about getting the "Non-global clients are not allowed access to APIv1" error from Auth0.

If we had a similar way to add custom form/body params like we do query string params, it would make life easier.

Cross-post side note: ideally the Swashbuckle.AspNetCore library could then add another extension method similar to [`OAuthAdditionalQueryStringParams`](https://github.com/domaindrivendev/Swashbuckle.AspNetCore/blob/c1fb708e3fe3f2c62abf4a85b91786f38b66582e/src/Swashbuckle.AspNetCore.SwaggerUI/SwaggerUIOptionsExtensions.cs#L244)  that would feed those vales through in a similar way.

**Comment by sferencik** [comment]
[Here](https://github.com/domaindrivendev/Swashbuckle.AspNetCore/issues/176#issuecomment-371277670) @karlssonsimon managed to supply `audience` in a query param, and apparently it worked. Sadly, it doesn't work for us - neither for our in-house auth endpoint, nor for an Auth0 endpoint we use. (Which is puzzling, since @karlssonsimon was also using Auth0.)

Even if the query-param trick worked, it would be fragile. The above is a great suggestion.

**Comment by Kizmar** [comment]
@sferencik If you review [actions.js](https://github.com/swagger-api/swagger-ui/blob/937c8f6208f3adf713b10a349a82a1b129bd0ffd/src/core/plugins/auth/actions.js), you can see what it sends based on the grant type.

If you are using the `implicit` grant type with Auth0; it's safe to add `audience` to the query string. In that case swagger-ui uses the ".../authorize" endpoint (which is the method used in the issue you linked). If you are using `client_credentials` or `password`, swagger-ui uses the ".../token" endpoint, which requires params in the body.

**Comment by sferencik** [comment]
Ah, that explains it. We're using `client_credentials` indeed.

**Comment by Kevweir** [comment]
I'm having the exact same problem using ASP>NET Core with the client_credentials flow.  There is no way to pass the audience value as part of the request.

**Comment by wdolek** [comment]
Hi @Kevweir, do you have (public) code/swagger config I could use for testing? We looked at it today with @sferencik and I think I have a plan.

Also big kudos to @Kizmar for describing the issue!

**Comment by sferencik** [comment]
:+1:

Looks promising. (I only checked the commits briefly from my phone.) Thanks for picking this up!

**Comment by sferencik** [comment]
I don't suppose this will display the audience in the UI form, will it?

**Comment by wdolek** [comment]
@sferencik I don't think so, but stay tuned, I'm not done yet :) (suggestions/help is welcomed)

**Comment by Kizmar** [comment]
It's been a while since I've dug into this, but in our case we use `appsettings.json` to hold Swagger (Swashbuckle) configurations applied in `Startup.cs` for deployment in different environments. I realize what I'm about to show is technically a different package (as mentioned in my OP), but it may help:

It would be nice if we could set the audience like this.
```
app.UseSwaggerUI(c =>
{
    c.OAuthClientId(options.ClientId);
    c.OAuthClientSecret(options.ClientSecret);
    c.OAuthAppName(options.AppName);
    c.OAuthAudience(options.Audience); // <-- Something like this
    ...
```

**Comment by wdolek** [comment]
@Kizmar, I'm aiming to latter one using additional params, similarly to additional query params - I would be worried adding new specialized method to library based on auth service (seems that Auth0 is doing something not-really-standard-ish). It is good compromise I think.

I will test my change and prepare PR if everything goes well. However I'm not JS/React/Redux developer, so there may be more things I'm not even seeing right now. 🤞

**Comment by wdolek** [comment]
Update: :disappointed: my approach does not work - `authSelectors.getConfigs()` is returning default state (empty), and not configured `securityDefinitions`. I will need to find other way of passing additional form paramaters.

Still working on it, help/guidance welcomed 🙇

**Comment by cpandya231** [comment]
I faced the same issue with Auth0. As @Kizmar  mentioned cant pass audience from swagger with client_credentials grant type. @wdolek I would like to help on this issue.

**Comment by wdolek** [comment]
@cpandya231 I have already created PR (#5724), but I tested my change against single scenario we do use. Also since I'm not React/Redux developer, my change may not be ideal. I'll be happy for another pair of eyes!

**Comment by Kizmar** [comment]
> ...Also since I'm not React/Redux developer, my change may not be ideal. I'll be happy for another pair of eyes!

Out of curiosity; where did React come up in all this? I am a React developer (supposedly), but my use of Swagger/Swashbuckle doesn't cross over into React-land. We use Swashbuckle with our .NET Core API's for testing purposes.

**Comment by cpandya231** [comment]
@wdolek @Kizmar I think the issue is with all the platform, weather it is React, .NET or Java. So the feature needs to be released for all the platforms.

**Comment by wdolek** [comment]
@Kizmar @cpandya231 My main struggle was to find place where actions are invoked and how configuration is read and propagated to correct place.

Second thing I wasn't sure about is why `additionalQueryStringParams` is taken from `authSelectors.getConfigs()` - same approach didn't work with newly introduced property and I had to grab it from `schema` (selectors contain initial object state rather than configurations supplied by user - I didn't find a way how to specify query string params trough UI, so I hope this is the right option for additional form params).

As I said, I managed to make it work as I needed, question is whether my change is _proper_ one or whether I missed something. Please do review. 🙇 

Getting additional query string params which I tried to mimic:
- [/src/core/oauth2-authorize.js:81](../blob/master/src/core/oauth2-authorize.js#L81) (this does not happen for my `flow`)
- [/src/core/plugins/auth/actions.js:154](../blob/master/src/core/plugins/auth/actions.js#L154)

**Comment by Asshiah** [comment]
Any updates on this issue? I am using Swagger-net (.NET package that provides Swagger-UI using Open Api 2.0 specification for .NET Full Framework 4.8) and I am still getting this issue where the audience parameter is missing in the request body for the token endpoint in order for auth0 to allow clientCredentials grant.

**Comment by wdolek** [comment]
@Asshiah Hello, looks like there's not enough interest in this feature. My PR #5724 got pretty much obsolete - after rebasing, it doesn't work any more and I won't have time to investigate what has changed, perhaps (hopefully) there's better way how to do it with current version. Feel free to look at it :) (or anyone else).

**Comment by vonWojtas** [comment]
Could we please do something about that? :D Would be great to be able to use `password` flow and add the `audience` in the request body parameter.
My setup:

Kotlin / SpringBoot 
Auth0

So actually I should ask in the spring community, but I got redirected here.
Any help? Thanks in advance!

**Comment by altescape** [comment]
It would be useful to add a way to be able to use Auth0's non-standard way.

In case it helps anyone we're currently getting around it using the requestInterceptor, here's an example:

```JavaScript
import SwaggerUI from 'swagger-ui'

const audience = 'me'

const ui = SwaggerUI({
  url: 'https://example.com/api.yaml',
  dom_id: '#swaggerUI',
  requestInterceptor: (request) => {
    if (request.url.includes('oauth/token')) {
      return {
        ...request,
        body: `${request.body}${audience}`
      }
    } else {
      return request
    }
  }
})
```

**Comment by ypicard** [comment]
This really needs to be implemented, the flexibility provided by this will certainly unlock so many people.

**Comment by simenflatby** [comment]
➕

**Comment by itsmaryann** [comment]
+1 We are unable to provide additional form data input from the UI so we cannot use the built in authorize.
